File: jeongyeon/swea_5215_햄버거 다이어트.c
#include <stdio.h>
#include <string.h>
#pragma warning(disable:4996)

int a[21][21];
int max, n, l;
void dfs(int idx, int score, int kal) {
	if (kal > l) {
		return;
	}
	if (idx == n) {
		if (score > max) max = score;
		return;
	}

	dfs(idx + 1, score + a[idx][0], kal + a[idx][1]);
	dfs(idx + 1, score, kal);
}

int main(void) {
	int t;
	scanf("%d", &t);
	for (int i = 1; i <= t; i++) {
		scanf("%d %d", &n, &l);
		for (int j = 0; j < n; j++) {
			for (int k = 0; k < 2; k++) {
				scanf("%d", &a[j][k]);
			}
		}
		max = 0;
		dfs(0, 0, 0);

		printf("#%d %d\n", i, max);
	}
	
}	

File: serim/week-03/swea_5215_햄버거다이어트.py
# swea 5215. 햄버거 다이어트 D3

# 민기가 매긴 햄버거 재료에 대한 점수
# 재료의 칼로리
# 둘다 보관하고 sum

T = int(input())

for tc in range(1, T+1):
    # 재료수, 제한칼로리
    N, L = map(int, input().split())

    ingredients = []

    for _ in range(N):
        score, calorie = map(int, input().split())
        ingredients.append((score, calorie))

    answer = 0  # 맛 점수 저장

    def dfs(idx, score_sum, calorie_sum):
        global answer

        if calorie_sum > L: # 칼로리제한을 넘어가면 return
            return

        if idx == N:
            answer = max(answer,score_sum)
            return

        score, calorie = ingredients[idx]
        # 선택하거나
        dfs(idx+1, score_sum+score, calorie_sum+calorie)
        # 선택안하거나
        dfs(idx+1, score_sum, calorie_sum)

    dfs(0, 0, 0)

    print(f"#{tc} {answer}")

        
